Implement full-on DSL→WDL translation

r-e-stern opened the pull request:
The DSL now supports parsing `.dsl` files to `.wdl`! The new `src/to_wdl.py` does this translation, restructuring the output of `parse.wdl` such that it matches the new `json`/`wdl` format. Its `generate_defaults()` methods generate default information required by Chiventure when it is not included in the `.dsl` file.

The primary changes here are that `to_wdl.py` is new, and the `dsl_parser.py` main function was moved to `parse.py`.

This pull request also includes: 
- **Four** example `.dsl` games whose `.wdl` output files run in Chiventure (see the updated `README.md` for more information).
- A modified file structure that gives each file type (`.dsl`, `.wdl`, `.lark`, `.py`) its own directory
- Updated documentation

Closes #938, closes #1076.

tessied commented on src/dsl/grammars/dsl_grammar.lark:
```diff
@@ -10,10 +10,10 @@ room: _NL "ROOM" id (property | connections)* item*
 action_ids: id ("," id)*
 
-property: _NL phrase "FOR" ("ITEM" | "ROOM") id ":" ESCAPED_STRING
+property: _NL phrase "FOR" id ":" ESCAPED_STRING -> misplaced_property
```
`property FOR object` is not supported; could remove this grammar rule.

jacknugent1529 replied:
While we ran out of time this quarter to fully implement the `property FOR object` syntax in a robust manner this quarter, I think it's something that should be on the future roadmap. Currently the `misplaced_property` function raises an error explaining that this syntax is not yet implemented. I think it would be best to leave the code as is, and add an issue to the backlog that next year's group can work on. That way, they will only have to deal with the transformer function instead of also rewriting the parsing aspect.

tessied replied:
Great! We should make a note of this in the documentation.

tessied commented on src/dsl/src/parse.py:
```diff
@@ -0,0 +1,46 @@
+from dsl_parser import export_dict
```
Might be helpful to add header comments to the python files to specify what they are for and what they contain.

jacknugent1529 replied:
Good idea. I added header comments in commit 3dcaeb6.

jacknugent1529 commented:
> For python 3.8.8, running `python src/parse.py <dsl file> <optional destination>` generates an error. @jacknugent1529 suggested adding `from __future__ import annotations` to the top of the python files on my end and that fixed the issue. Consider adding this to all the python files so earlier versions of python can convert a DSL file to a WDL file.

It turns out a few specific type hints we used were not introduced until python 3.9. I added the `from __future__ import annotations` to each file, which provides compatibility with python 3.7 and later. The readme reflects this change. I also added a `conda_environment.yml` file which makes it easier to set up the environment and specifies python 3.9. At this point, I have tested that the program works for all python >3.7, but it is best to use python 3.9.

jacknugent1529 commented:
I just made a few more changes to add improved flags to the interface. The `--debug` flag now has several more options to debug various stages. Additionally, there is a `--no-write` flag to not write to a file and a `--ignore-warnings` flag if the user does not wish to view the warnings.

tessied commented:
`short_example.dsl`, `long-example.dsl`, and `vars.dsl` cannot be loaded into Chiventure. We should either remove these files or make a note of this in the `README` file.

r-e-stern commented:
> `short_example.dsl`, `long-example.dsl`, and `vars.dsl` cannot be loaded into Chiventure. We should either remove these files or make a note about this in the `README` file.

The section of the `README` which notes this does it in a sort of roundabout way: the files which load in Chiventure are listed in bold type. The `README` notes this:

> ###### While all example `.dsl` files compile to `.wdl`, only some resulting files (the bold ones) will run in Chiventure; the others are either not complete games, contain features not supported by WDL/Chiventure, or both.

We could try to make this more clear if you think it's too confusing.

tessied commented:
> > `short_example.dsl`, `long-example.dsl`, and `vars.dsl` cannot be loaded into Chiventure. We should either remove these files or make a note about this in the `README` file.
> 
> The section of the `README` which notes this does it in a sort of roundabout way: the files which load in Chiventure are listed in bold type. The `README` notes this:
> 
> > ###### While all example `.dsl` files compile to `.wdl`, only some resulting files (the bold ones) will run in Chiventure; the others are either not complete games, contain features not supported by WDL/Chiventure, or both.
> 
> We could try to make this more clear if you think it's too confusing.

I must've missed the difference between the bold ones and the regular ones the first time I looked at it, but I think your explanation is good! If we want to be more clear about it, I was thinking maybe we could make a note about this in the description of each file.

tessied reviewed (approved):
The files are nicely organized and the example dsl files can be successfully parsed to wdl and then loaded into chiventure. Looks good!

ecoble commented on src/dsl/examples/dsl/long-example.dsl:
```diff
@@ -0,0 +1,41 @@
+    
+# testing: generating room short description by default, 'long description' synonym
+ROOM prison hallway
+    long description: "The prison hallway is dark and foreboding, but hell, you knew that already, you live here. Honestly, you're kind of a fan of Hallway B (that's its official name): even though it's dark and brutalistr and home to a rat colony, it' isn't Hallway C. Everyone knows Hallway C is haunted, and you'd sooner die than have to live there. Not that it matters, rumor has it, if you move ...
```
Typo in brutalist, and looks like an extra apostrophe after it.

r-e-stern replied:
Fixed.

ecoble commented on src/dsl/examples/dsl/vars.dsl:
```diff
@@ -0,0 +1,27 @@
+# Define variables
```
This file looks like a duplicate of one that was just merged in the other PR, watch out for that when updating this branch.

ecoble commented on src/dsl/examples/wdl/long-example.wdl:
```diff
@@ -0,0 +1,85 @@
+  },
+  "ROOMS": {
+    "prison hallway": {
+      "long description": "The prison hallway is dark and foreboding, but hell, you knew that already, you live here. Honestly, you're kind of a fan of Hallway B (that's its official name): even though it's dark and brutalistr and home to a rat colony, it' isn't Hallway C. Everyone knows Hallway C is haunted, and you'd sooner die than have to live there. Not that it matters, rumor has it, if you  ...
```
Same typos in here

r-e-stern replied:
Generated a new version that fixes this too.

ecoble commented on src/dsl/grammars/vars.lark:
```diff
@@ -0,0 +1,39 @@
+?start: (assignment | reference | other)* -> start
```
This file also looks like its duplicated from the other PR

ecoble commented on src/dsl/src/vars_parser.py:
```diff
@@ -0,0 +1,149 @@
+"""This module evaluates references to variables in a dsl file."""
```
Same on duplicating this file.

ecoble reviewed (changes requested):
This looks great, and your sample games are really funny! A couple small typos I saw, and also be careful when merging the new changes from your last PR, it looks like a few files are duplicates in new locations so you might have to go back and move some stuff around.

r-e-stern commented:
What do we do to solve the duplicates problem? We want to keep our new file structure. Since the `dsl/to-wdl` branch is up-to-date with every feature we've introduced so far, at merge, I think we just want to replace the entire directory.
```
OLD                                          NEW
├── examples (includes .dsl files only)      ├── conda_environment.yml
├── README.md                                ├── examples
├── src (includes .py and .lark files)       │   ├── dsl (.dsl files)
└── ui_debug.txt                             │   └── wdl (.wdl files)
                                             ├── grammars (.lark files)
                                             ├── README.md
                                             ├── src (.py files)
                                             └── ui_debug.txt
```

r-e-stern commented:
After doing some testing, I don't think the 'duplicate files' thing will end up causing a problem. We didn't create the files in parallel, we just merged the `dsl/vars` branch into this one, so Git will have a record of the relevant commits on hand. I think the merge will go smoothly.

To test this, I created a remote branch `dsl/test-for-merging` with exactly the same contents as `dev`, and attempted to merge this branch into it a couple times. The merge always went smoothly. Here's a log of what happened: 
```
(base) robertestern@linux6:~/cs220/chiventure/src/dsl$ git checkout dev
Switched to branch 'dev'
Your branch is up to date with 'origin/dev'.
(base) robertestern@linux6:~/cs220/chiventure/src/dsl$ tree
.
├── examples
│   ├── distaff.dsl
│   ├── short_example.dsl
│   └── vars.dsl
├── README.md
└── src
    ├── dsl_grammar.lark
    ├── dsl_parser.py
    ├── util.lark
    ├── vars.lark
    └── vars_parser.py

2 directories, 9 files
(base) robertestern@linux6:~/cs220/chiventure/src/dsl$ git checkout -b dsl/test-for-merging
Switched to a new branch 'dsl/test-for-merging'
(base) robertestern@linux6:~/cs220/chiventure/src/dsl$ git merge dsl/to-wdl
Merge made by the 'recursive' strategy.
 src/dsl/README.md                            |  90 +++++++++++++++++++++---
 src/dsl/conda_environment.yml                |   6 ++
 src/dsl/examples/dsl/connected-rooms.dsl     |  27 +++++++
 src/dsl/examples/dsl/connected-vars.dsl      |  11 +++
 src/dsl/examples/{ => dsl}/distaff.dsl       |   3 +-
 src/dsl/examples/dsl/long-example.dsl        |  41 +++++++++++
 src/dsl/examples/dsl/min.dsl                 |  15 ++++
 src/dsl/examples/{ => dsl}/short_example.dsl |   4 +-
 src/dsl/examples/{ => dsl}/vars.dsl          |   0
 src/dsl/examples/wdl/connected-rooms.wdl     |  65 +++++++++++++++++
 src/dsl/examples/wdl/connected-vars.wdl      |  70 +++++++++++++++++++
 src/dsl/examples/wdl/distaff.wdl             |  75 ++++++++++++++++++++
 src/dsl/examples/wdl/long-example.wdl        |  85 +++++++++++++++++++++++
 src/dsl/examples/wdl/min.wdl                 |  82 ++++++++++++++++++++++
 src/dsl/examples/wdl/short_example.wdl       |  44 ++++++++++++
 src/dsl/examples/wdl/vars.wdl                |  41 +++++++++++
 src/dsl/{src => grammars}/dsl_grammar.lark   |   4 +-
 src/dsl/{src => grammars}/util.lark          |   3 +-
 src/dsl/{src => grammars}/vars.lark          |   4 +-
 src/dsl/src/dsl_parser.py                    |  78 ++++++++++++---------
 src/dsl/src/parse.py                         | 111 +++++++++++++++++++++++++++++
 src/dsl/src/to_wdl.py                        | 274 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 src/dsl/src/vars_parser.py                   |  21 ++++--
 23 files changed, 1095 insertions(+), 59 deletions(-)
 create mode 100644 src/dsl/conda_environment.yml
 create mode 100644 src/dsl/examples/dsl/connected-rooms.dsl
 create mode 100644 src/dsl/examples/dsl/connected-vars.dsl
 rename src/dsl/examples/{ => dsl}/distaff.dsl (98%)
 create mode 100644 src/dsl/examples/dsl/long-example.dsl
 create mode 100644 src/dsl/examples/dsl/min.dsl
 rename src/dsl/examples/{ => dsl}/short_example.dsl (87%)
 rename src/dsl/examples/{ => dsl}/vars.dsl (100%)
 create mode 100644 src/dsl/examples/wdl/connected-rooms.wdl
 create mode 100644 src/dsl/examples/wdl/connected-vars.wdl
 create mode 100644 src/dsl/examples/wdl/distaff.wdl
 create mode 100644 src/dsl/examples/wdl/long-example.wdl
 create mode 100644 src/dsl/examples/wdl/min.wdl
 create mode 100644 src/dsl/examples/wdl/short_example.wdl
 create mode 100644 src/dsl/examples/wdl/vars.wdl
 rename src/dsl/{src => grammars}/dsl_grammar.lark (84%)
 rename src/dsl/{src => grammars}/util.lark (78%)
 rename src/dsl/{src => grammars}/vars.lark (85%)
 create mode 100644 src/dsl/src/parse.py
 create mode 100644 src/dsl/src/to_wdl.py
(base) robertestern@linux6:~/cs220/chiventure/src/dsl$ tree
.
├── conda_environment.yml
├── examples
│   ├── dsl
│   │   ├── connected-rooms.dsl
│   │   ├── connected-vars.dsl
│   │   ├── distaff.dsl
│   │   ├── long-example.dsl
│   │   ├── min.dsl
│   │   ├── short_example.dsl
│   │   └── vars.dsl
│   └── wdl
│       ├── connected-rooms.wdl
│       ├── connected-vars.wdl
│       ├── distaff.wdl
│       ├── long-example.wdl
│       ├── min.wdl
│       ├── short_example.wdl
│       └── vars.wdl
├── grammars
│   ├── dsl_grammar.lark
│   ├── util.lark
│   └── vars.lark
├── README.md
└── src
    ├── dsl_parser.py
    ├── parse.py
    ├── to_wdl.py
    └── vars_parser.py

5 directories, 23 files
```

ecoble reviewed (approved):
LGTM! As long as you are aware about the files, it is okay, I just wanted to bring it to your attention.

r-e-stern commented:
Closing statement: We added the last step of `.dsl`-to-`.wdl` translation into the dsl by enabling it to convert from the intermediate parse-tree state outputted by `dsl_parser.py` into the established WDL/JSON format. Most of the relevant changes were in the new `to_wdl.py` file which takes `Room`, `Item`, and `Game` Python objects as input and converts them to JSON, though some documentation, example games, and `conda` shenanigans were also included. Since these changes have been successfully merged, this closes issue #938 and issue #1076.

ecoble commented:
Issue Score: ✔️++

Comments:
Excellent work! Everyone did a really great job keeping this PR updated, and doing a really thorough review of the code. 

<!-- {"score": "check-plus-plus"} -->
